### DISCUSSION TITLE
DLL warnings when building under x86-mingw-* community triplets. Should they be ignored or addressed?

### DISCUSSION THREAD

**Discussion by XJ-0461** [microsoft/vcpkg#39961] [Q&A] [answered]
When cross-compiling updated Lua port #39953 with the community x86-mingw-* triplets and a Linux MinGW toolchain. I produce a liblua.dll and a liblua-c++.dll as expected. However, a few warnings are reported:

```
/.../my-port/lua/portfile.cmake: warning: Import libraries for installed DLLs appear to be missing. If this is intended, add set(VCPKG_POLICY_DLLS_WITHOUT_LIBS enabled)
/.../my-port/lua/portfile.cmake: warning: DLLs that link with obsolete C RunTime ("CRT") DLLs were installed. Installed DLLs should link with an in-support CRT. You can inspect the dependencies of a DLL with `dumpbin.exe /dependents mylibfile.dll`. If you're using a custom triplet targeting an old CRT, add set(VCPKG_POLICY_ALLOW_OBSOLETE_MSVCRT enabled) to the triplet's .cmake file. To suppress this message for this port, add set(VCPKG_POLICY_ALLOW_OBSOLETE_MSVCRT enabled)
/.../build/external/vcpkg/packages/lua_cool-vcpkg-custom-triplet: note: the DLLs are relative to ${CURRENT_PACKAGES_DIR} here
note: debug/bin/liblua.dll
note: debug/bin/liblua-c++.dll
note: bin/liblua.dll
note: bin/liblua-c++.dll
```

When inspecting them I see these dependencies:

```
objdump -x liblua.dll | grep "DLL Name:"
        DLL Name: libgcc_s_dw2-1.dll
        DLL Name: KERNEL32.dll
        DLL Name: msvcrt.dll
```
```
objdump -x liblua-c++.dll | grep "DLL Name:"
        DLL Name: libgcc_s_dw2-1.dll
        DLL Name: KERNEL32.dll
        DLL Name: msvcrt.dll
        DLL Name: libstdc++-6.dll
```

1. These dependencies dont look like they should be distributed with the library itself. To me, these look like they should be found by the system at runtime.
  - If I'm wrong about that assumption, what is the preferred way to copy these over?

2. Is there any way to "link with an in-support CRT." while using MinGW and the 32-bit toolchains?
  - Or is this a case where we should disable that warning?

**Comment by dg0yt** [comment, marked as answer]
- The DLLs are required to be found at runtime. kernel32.dll and msvcrt.dll can be assumed to be provided by Windows. libgcc\* and libstdc++\* must be deployed to the target system either with the app or with the same mingw installation. IMO the preferred way is to copy them along with your app, to ensure compatible versions. However, you should study the implications of the relevant licenses.

- To link with a supported CRT, choose a suitable mingw environment. Msys2 (for Windows hosts) has UCRT environment for i686, x86_64 and arm64.

**Reply by XJ-0461** [reply]
- I'm surprised that the preferred way is to copy over the standard libraries. Because it seems like something that you can reasonably expect the target system to have already.

- I see, so continue to use the \*-mingw-\* triplets as before however, cross-compile with a toolchain for example:
```cmake
set(CMAKE_SYSTEM_NAME Windows)
set(CMAKE_HOST_SYSTEM_NAME Linux)

set(CMAKE_CROSSCOMPILING ON)
set(CMAKE_CROSSCOMPILING_EMULATOR /usr/bin/wine)

set(CMAKE_C_COMPILER /usr/bin/x86_64-w64-mingw32ucrt-gcc CACHE STRING INTERNAL)
set(CMAKE_CXX_COMPILER /usr/bin/x86_64-w64-mingw32ucrt-g++ CACHE STRING INTERNAL)
set(CMAKE_FIND_ROOT_PATH /usr/x86_64-w64-mingw32ucrt)

```

**Reply by dg0yt** [reply]
> * I'm surprised that the preferred way is to copy over the standard libraries. Because it seems like something that you can reasonably expect the target system to have already.

You are *not* expected to copy libraries "that you can reasonably expect the target system to have already." I explicitly named the libs which can be expected (part of Windows) and which cannot reasonably be expected (part of MinGW/GCC). 
And note that even UCRT cannot be expected on older Windows systems.
